### ISSUE TITLE
Explore Notebook Level formatting support + api

### ISSUE THREAD

**Issue by Yoyokrazy** [microsoft/vscode#207334]
Tracking issue to follow progress for native notebookk formatting support within VS Code. Will need proposed api alongside it, as well as api commonds on the EH side.

**Comment by Yoyokrazy** [comment]
Currently we have the following two commands:
- `notebook.format`
https://github.com/microsoft/vscode/blob/a1815b994a820e85aff5c0dba85bdcee72c0b8f5/src/vs/workbench/contrib/notebook/browser/contrib/format/formatting.ts#L37-L108

- `notebook.formatCell`
https://github.com/microsoft/vscode/blob/a1815b994a820e85aff5c0dba85bdcee72c0b8f5/src/vs/workbench/contrib/notebook/browser/contrib/format/formatting.ts#L111-L137

Both of these actions take an approach that doesn't really include the context of the entire notebook at once. The former takes a naive approach and simply calls down to `provideDocumentFormattingEdits()` for every cell in parallel. The latter does the same, albeit just for a single cell. 

The ask here is to create api and infrastructure to be able to format an entire notebook at once with the context of what is in each cell. The current limitation is that `provideDocumentFormattingEdits()` only takes in a single `TextDocument` and returns a `TextEdit[]`. An example use-case where this is slightly limiting is when you attempt to design a formatter that might shift all imports to a central first cell, or remove duplicated import statements. With the current approach, formatting every cell in parallel, there isn't a good solution to this as when formatting a single cell you wouldn't be aware of imports in other cells. 

The proposal is to rename the above commands into something that more resembles `notebook.formatAllCells` +`notebook.formatSingleCell`, then introduce a `notebook.format` command that will leverage the following new api:

```ts
declare module 'vscode' {

	export interface NotebookFormattingEditProvider {

		/**
		 * Provide formatting edits for a whole notebook.
		 *
		 * @param notebook The NotebookDocument in which the command was invoked.
		 * @param options Options controlling formatting.
		 * @param token A cancellation token.
		 * @returns A set of Workspace edits or a thenable that resolves to such. The lack of a result can be
		 * signaled by returning `undefined`, `null`, or an empty array.
		 */
		provideNotebookFormattingEdits(notebook: NotebookDocument, options: FormattingOptions, token: CancellationToken): ProviderResult<WorkspaceEdit[]>;
	}
}
```

**Comment by Yoyokrazy** [comment]
Feedback from API sync (cc @rebornix):
- Seems slightly closer to a definition problem, what is a formatter, what is a CodeAction
  - Formatter => add or remove whitespace, newlines, etc. 
  - CodeAction => anything that is adding or removing characters (semicolons, copyrights, spelling, organizing imports, etc.)
- Not really a need for new api here as it can be accomplished with a `notebook.source.xyz` type codeaction
  - One of the most common asks is for a formatter that can extract and shift all import statements to a central top cell. 
  - While it may not be performant, polished, or finished, I've already played around with the [idea here](https://github.com/Yoyokrazy/clean-nb-imports-ext).
- LSP limitations (cc @karthiknadig) related to LSP not knowing about notebook edits (need to revist this to understand how all the pieces fit together) which is an argument in favor of a notebook formatting provider maybe
- What tooling limitations do Ruff/Black have that prevents them from creating codeactions to accomplish these formatting needs?

**Comment by Yoyokrazy** [comment]
Closing in favor of #212381
